libgraph/src/analytics/k_shortest_paths/ksssp.cpp

```diff
@@ -0,0 +1,464 @@
+/*
+ * This file belongs to the Galois project, a C++ library for exploiting
+ * parallelism. The code is being released under the terms of the 3-Clause BSD
+ * License (a copy is located in LICENSE.txt at the top-level directory).
+ *
+ * Copyright (C) 2018, The University of Texas at Austin. All rights reserved.
+ * UNIVERSITY EXPRESSLY DISCLAIMS ANY AND ALL WARRANTIES CONCERNING THIS
+ * SOFTWARE AND DOCUMENTATION, INCLUDING ANY WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR ANY PARTICULAR PURPOSE, NON-INFRINGEMENT AND WARRANTIES OF
+ * PERFORMANCE, AND ANY WARRANTY THAT MIGHT OTHERWISE ARISE FROM COURSE OF
+ * DEALING OR USAGE OF TRADE.  NO WARRANTY IS EITHER EXPRESS OR IMPLIED WITH
+ * RESPECT TO THE USE OF THE SOFTWARE OR DOCUMENTATION. Under no circumstances
+ * shall University be liable for incidental, special, indirect, direct or
+ * consequential damages or loss of profits, interruption of business, or
+ * related expenses which may arise from use of Software or Documentation,
+ * including but not limited to those resulting from defects in Software and/or
+ * Documentation, or loss or inaccuracy of data of any kind.
+ */
+
+#include "katana/analytics/k_shortest_paths/ksssp.h"
+
+#include "katana/AtomicHelpers.h"
+#include "katana/Reduction.h"
+#include "katana/Statistics.h"
+#include "katana/TypedPropertyGraph.h"
+#include "katana/analytics/BfsSsspImplementationBase.h"
+#include "katana/analytics/kSsspImplementationBase.h"
+#include "katana/gstl.h"
+
+using namespace katana::analytics;
+
+typedef uint32_t Parent;
+
+struct Path {
+  Parent parent;
+  const Path* last{nullptr};
+};
+
+struct NodeCount : public katana::AtomicPODProperty<uint32_t> {};
+
+struct NodeMax : public katana::AtomicPODProperty<uint32_t> {};
+
+using EdgeWeight = katana::UInt32Property;
+
+using NodeData = std::tuple<NodeCount, NodeMax>;
+using EdgeData = std::tuple<EdgeWeight>;
+using Weight = uint32_t;
+using Distance = uint32_t;
+
```

Comment:
oh I just realized that we need to use Weight as a template parameter, so as to allow all type of edge weights. both sssp and louvain has this functionality. see this: https://github.com/uditagar/katana/blob/master/libgraph/src/analytics/sssp/sssp.cpp

Reply:
So all values with fixed weight need to be changed with template weight?

Reply:
yes